### ISSUE TITLE
App Analytics dashboards

### ISSUE THREAD

**Issue by anirudha** [opensearch-project/observability#131]
### Overview
Application Analytics allows OpenSearch Dashboards users to select logs, traces, and metrics to be part of an application that can be monitored for overall health and visualized on one page. It also allows users to quickly pivot between traces, metrics, and logs to dig into the source of any issues that arise.

### Problem
Currently, users have no way of identifying which trace and log data belongs to which application because all trace data ends up in the same index and log data goes into different independent indices. This application is made up of user-selected logs, metrics, and traces. The current solution forces users to sift through large volumes of data to get to what they really need. Also this data, distributed across Trace Analytics, Event Analytics, Dashboards, does not provide a comprehensive view of an application’s health at a glance. 

#### Home Page
1. Users are able to view existing applications and some high level information (Name, Composition, Current Availability, Availability Metrics)
2. Users are able to create/rename/duplicate/delete applications

##### Application Information
1. Users are able to specify name of new application
2. Users are able to optionally provide a description of their application

##### Composition
1. Users are able to configure base query
2. Users are able to select services and entities to include in the application
3. Users are able to constrain application to specific trace groups

##### Availability
1. Users are able to set custom levels of availability according to query conditions

##### Overview Tab
1. Users are able to view Latency by Trace Group
2. Users are able to view default charts: Application Composition Map,  Error Rate, Throughput
3. Users are able to add new charts

##### Services Tab
1. Users are able to view information about specific or all services in a table

##### Traces & Spans Tab
1. Users are able to view general information about traces and spans in two tables
2. Users are able to view more detailed information by selecting a trace or span

##### Log Events
1. Users are able to query and analyze logs for their application
2. Users are able to create visualizations that are saved as metrics

##### <del>Panels</del> Metrics
1. Users are able to view metrics (visualizations) created from the log events tab

##### Configuration
1. Users are able to edit their application composition and availability settings

### UX/UI Design
[Figma Link](https://www.figma.com/proto/XPM1dDghgqlZll3QQ21FAa/Observability-GA_3rd-iter-(Latest)?page-id=958%3A65045&node-id=998%3A65686&viewport=369%2C48%2C0.17&scaling=min-zoom&starting-point-node-id=1049%3A83712)
Home Page
![Screen Shot 2021-11-17 at 9 37 52 AM](https://user-images.githubusercontent.com/92330893/143332252-21ca1324-f4d3-4a84-9eaa-e5fb4782216c.png)

Overview Tab
![overview](https://user-images.githubusercontent.com/92330893/143332257-57c4b6a0-ce80-4132-8af0-763437adf5e0.gif)

Services Tab
![services](https://user-images.githubusercontent.com/92330893/143332261-7b9bfb20-17be-45b4-b222-cb6c70203af7.gif)

Configuration Tab
![config](https://user-images.githubusercontent.com/92330893/143332282-8004c827-74e0-4fea-82f7-e8667f1c2575.gif)

#### Home Page
- [x] Add table to view existing applications
- - [x] #376
- - [x] #340 
- [x] #489
- [x] Add a button to create a new application
- - [x] Redirect to Create Page if clicked
- [x] #326
- - [x] #327
- - [x] #328 
- - [x] #343 
- [x] #378

#### Create Page
- [x] Add create button
- - [x] #319
- - [x] #329
- - [x] #323
- - [x] #324
- - [x] #377 
- [x] Add cancel button
- - [x] #325
- [x] #418
##### Application Information
- [x] Add text form fields for name(required) and description(optional)
- - [x] #288 
##### Composition
- [x] Add accordion for Log Sources
- - [x] #292 
- - [x] #289 
- - [x] #290 
- - [x] #291 
- - [x] #305 
   - - [x] #320 
- [x] Add accordion for Services & entities
- - [x] Add select form field where users can select services and entities
- - [x] #293 
- - [x] #304 
   - - [x] #321 
- [x] Add accordion for Trace groups
- - [x] #295 
- - [x] #296 
- - [x] #308
   - - [x] #322 
#### Application Details
- [x] #382
- [x] #387
- [x] #456
- [x] #457
- [x] #458
- [x] #459
##### Overview Tab (Dashboard from Trace Analytics)
- [x] #384
- [x] #388
- [x] Add table to view latency of trace groups
- - [x] Add filter for < 95th percentile or >= 95th percentile 
- - [x] Add pop up for enlarged view of 24-hour latency trend
- - [x] Make Traces column data value a link
  - - [x] Redirect to Traces & Spans tab with filter of trace group
- - [x] Make Trace group name column data value a link
- [x] Add Application Composition Map chart
- - [x] #389 
- [x] Add Error Rate chart
- [x] Add Throughput chart
- [ ] #460
##### Services Tab (Services from Trace Analytics)
- [x] #390
- [x] Add filter bar
- - [x] #391
- - [x] #392
- [x] Add table to view services
- - [x] Make Name column data value a link
  - - [x] #393
- - [x] Make Traces column data value a link
  - - [x] #394
##### Traces & Spans Tab (Traces and Span Details Table from Trace Analytics)
- [x] #309
- - [x] Filter traces and spans by trace id or trace group name
- - [x] #395 
- [x] Add table to view Traces
- - [x] Make Trace ID column a link
  - - [x] #396 
- [x] #397
- - [x] #398 
  - - [x] #399 
    - - [x] #400 
- [x] Add table to view Spans
- - [x] Make Span ID column a link
  - - [x] #401 
    - - [x] #402 
##### Log Events (Log Explorer from Event Analytics)
- [x] Remove query tabs
- [x] Remove save button for events
- [x] Add PPL Query search bar
- - [x] #429 
- - [x] #430 
- - [x] #431 
- [x] Save visualization to panels tab
#####  <del>Panels</del> Metrics (Custom Panel View from Operational Panels)
- [x] #412
- [x] #385
- [x] #386 
- [x] Add Add Visualization Button
- - [x] #432 
- [x] #416
- [x] #461
- [x] #463
- [x] #464
- [x] #465
- [x] #466
##### Configuration
- [x] #433
- [x] #434
- - [x] #435 
- - [x] #467 
- [x] #436
- [x] #437
##### Backend
- [x] #403
- [x] #409
- [x] #410
- [x] #411
##### Known Bugs
- [-] #468 
- [-] #469 
- [-] #470 
- [-] #471 
- [x] #484
- [x] #485
- [x] #486
- [x] #487
- [x] #488 
- [x] #499
- [x] #537

#### Relationship to Other Modules

[Dashboard](https://github.com/opensearch-project/trace-analytics/blob/main/dashboards-observability/public/components/trace_analytics/components/dashboard/dashboard.tsx), [Trace](https://github.com/opensearch-project/trace-analytics/blob/main/dashboards-observability/public/components/trace_analytics/components/traces/traces.tsx), [Span Detail](https://github.com/opensearch-project/trace-analytics/blob/main/dashboards-observability/public/components/trace_analytics/components/traces/span_detail_panel.tsx), [Service](https://github.com/opensearch-project/trace-analytics/blob/main/dashboards-observability/public/components/trace_analytics/components/services/services.tsx) from Trace Analytics, [Log Explorer](https://github.com/opensearch-project/trace-analytics/blob/main/dashboards-observability/public/components/explorer/log_explorer.tsx) from Event Analytics, and [Panels](https://github.com/opensearch-project/observability/blob/main/dashboards-observability/public/components/custom_panels/custom_panel_view.tsx) from Operational Panels will be re-used in Application Analytics. We will need to refactor the components to become more modular. We also need to make sure only the selected services are filtered through when displaying information.

![Architecture Diagram](https://user-images.githubusercontent.com/92330893/145495954-4ea2b2ed-c85a-4b0d-9c36-c947e9c4d7aa.png)

##### Application

- **name:** User-defined name of application
- **description (optional):** User-defined description of the application
- **baseQuery:** PPL query to filter the application by
- **servicesEntities:** List of services and entities included in application
- **traceGroups:** List of trace groups included in application
( - **availabilityLevels:** User-defined thresholds that signify custom levels of availability of application
    - **label**
    - **levelPriority**
    - **color**
    - **condition** ) Temporarily not added

**Model**
```
{
   "properties": {
        "name": {
            "type": "keyword",
            },
        "description": {
            "type": "text",
            },
        "baseQuery": {
            "type": "text",
            },
        "servicesEntities": {
            "type": "nested",
        },
        "traceGroups": {
            "type": "nested",
        },
    }
}
```

**Example**
```
{
  "properties": {
    "name": "Application 1",
    "description": "This is my app",
    "baseQuery": "source = flight_logs",
    "servicesEntities": ["Payment"],
    "traceGroups" : ["Payment.auto"]
    }
}
```
#### Options for Structure of Availability Levels

**Option 1** - List of objects with all of the fields
Pros: Simple
```
"availabilityLevels": [
    {
    "label": "Available",
    "color": "#50C878",
    "condition": "If otherwise",
    "levelPriority": 0,
    }, {
    "label": "Unavailable",
    "color": "#FF0000",
    "condition": "when errorRate() >= 2",
    "levelPriority": 1,
    },
]
```
**Option 2** - List of objects with key/value pairs. (key: levelPriority) (value: label, color condition in object)
Pros: Easier to sort, 
```
"availabilityLevels": [
    0: {
    "content": {
        "label": "Available",
        "color": "#50C878",
        "condition": "If otherwise",
        }
    }, 
    1: {
    "content": {
        "label": "Unavailable",
        "color": "#FF0000",
        "condition": "when errorRate() >= 2",
        }
    },
]
```
#### API Design

`APP_ANALYTICS_API_PREFIX='/api/observability/application'`

* Fetch all Applications
```
GET ${APP_ANALYTICS_API_PREFIX}/

RESPONSE BODY
{
  "statusCode": 200,  
  "message": "Ready", 
  "applications": [
    {
      "id": "2FF3GW3H8",
      "name": "Application 1",
      "servicesEntities": ["Payment", "Users"],
      "traceGroups" : ["Payment.auto", "Users.admin"]
    },
    {
      "id": "2FHEP953H",
      "name": "Application 2",
      "servicesEntities": ["Purchase"],
      "traceGroups" : ["Purchase.source"]
    }
  ]
}  
```
* Fetch Application by ID
```
GET ${APP_ANALYTICS_API_PREFIX}/{application_id} 
    
    RESPONSE BODY
    {
      "statusCode": 200, 
      "message": "Application Fetched", 
      "application": {
            "name": "Application 1",
            "description": "Description for the application",
            "base_query": "source = opensearch_sample_database_flights",
            "services_entities": [
                "Payment",
                "Users",
                "Purchase"
            ]
            "trace_groups": [
                "Payment.auto",
                "Users.admin",
                "Purchase.source",
            ]
        }
    }
```
* Create Application
```
POST ${APP_ANALYTICS_API_PREFIX}/
    
    REQUEST BODY
    {
       "name": "Application 1",
        "description": "Description for the application",
        "base_query": "source = opensearch_sample_database_flights",
        "services_entities": [
            "Payment",
            "Users",
            "Purchase"
        ]
        "trace_groups": [
            "Payment.auto",
            "Users.admin",
            "Purchase.source",
        ]
    }
    
    RESPONSE BODY
    {
      "statusCode": 200, 
      "message": "Application Created", 
      "newAppId": "2FG6FWGY5" // New application ID
    }
```
* Update Application - all fields are updatable
```
PATCH ${APP_ANALYTICS_API_PREFIX}/{application_id} 
    
    REQUEST BODY
    {
      "appId": "2FG6FWGY5", // application id to be renamed 
      "base_query": "source = opensearch_sample_database_flights" // fields to update
    } 
    
    RESPONSE BODY
    {
      "statusCode": 200, 
      "message": "Application Updated"
    }
```
* Duplicate Application
```
POST ${APP_ANALYTICS_API_PREFIX}/clone 
    
    REQUEST BODY
    {
      "appId": "2FG6FWGY5", // application id to be renamed 
      "name":"Application 1 (copy)" // new name for the application to be duplicated 
    } 
    
    RESPONSE BODY
    {
      "statusCode": 200, 
      "message": "Application Cloned",  
      "newAppId": "2FFAG7HAQ" // newly duplicated application id  
    }
```
* Delete Application
```
DELETE ${APP_ANALYTICS_API_PREFIX}/{application_ids}
    
    RESPONSE BODY
    {
      "statusCode": 204, 
      "message": "Application Deleted",   
    }
```

#### Autocomplete Logic
Possible [PPL commands](https://opensearch.org/docs/1.2/observability-plugins/ppl/commands/): source, dedup, eval, fields, head, rare, rename, sort, stats, top, where

##### for Base Query on Create Page
- Not allowed commands: stats, head, rare, top
- ex:   source = opensearch_dashboard_sample_flights | where DestCityName = 'Venice' 

##### ~~for Availability Level Condition on Create Page~~ (Conditions can not use PPL query)
- ~~Allowed commands: where, eval, rename~~
- ~~Condition must be evaluate to boolean~~

##### for Filter on Log Events Tab
- Not allowed commands: source
- ex:  where OriginCityName = 'Rome' | stats avg( AvgTicketPrice) 

##### ~~for Filter on Traces & Spans Tab (pseudo)~~ (Using filter instead of PPL query)

**Comment by ryn9** [comment]
@anirudha 
Are these solution patterns being developed with multi-tenancy and various RBAC in mind?
Are there issues and/or other docs that could be linked back to on these solution patterns?

**Comment by anirudha** [comment]
yes, the observabilty plugin supports RBAC  and multi -tenancy using the security plugin

**Comment by ryn9** [comment]
@anirudha additionally will the plugin support customizable index patterns, etc.. ?
The trace analytics plug-in required use of otel-v1-apm-span-* and otel-v1-apm-service-map*, which will not work in all environments, especially when multi-tenancy comes into play.

**Comment by eugenesk24** [comment]
@ryn9 Hi Ryan, I talked to Ani about your question. We are planning on supporting regex based index matching in PPL soon, I think he said next release. As to your second question, I'm a little confused how multi-tenancy will affect multiple index matching? I think the security plugin should handle any access restrictions for indices. Please feel free to elaborate.

**Comment by ryn9** [comment]
@eugenesk24 
>>As to your second question, I'm a little confused how multi-tenancy will affect multiple index matching? 
I was trying to say that making specific index patterns required for plugin use can have impact on what is, probably, the most common way to handle multi-tenancy + RBAC - role based access to specific indices. So - I am hoping that static index names are not being considered like they were with the initial trace analytics plugin.

To note - plugins haven't always been thought with multi-tenancy + RBAC in mind - so I am hoping to find out more details about how the observability plugin is being designed with this in mind.

**Comment by anirudha** [comment]
Hi @ryn9 ,
Could you please elaborate on a use-case you are looking for in some details, for now here are the details.
- Observability plugin support same tenancy structures as all of dashboards eg. [link](https://opensearch.org/docs/latest/security-plugin/access-control/multi-tenancy/)
- RBAC is i.e roles based access control will follow the rules configured by your custer admin on the opensearch setup.
- Currently we don't support index patterns for wildcard indices, this will be a supported in the next few releases. The wildcards will match any permission-ed index available to user for this roles.

> The trace analytics plug-in required use of otel-v1-apm-span-* and otel-v1-apm-service-map*, which will not work in all environments, especially when multi-tenancy comes into play.
could you explain this with an example. From what i know access to these are only based on your roles and not tenancy.

**Comment by ryn9** [comment]
>>could you explain this with an example. From what i know access to these are only based on your roles and not tenancy.
Sorry to confuse - 'tenancy' was probably the wrong word here.

The trace analytics plug-in required use of otel-v1-apm-span-* and otel-v1-apm-service-map*, which will not work in all environments. For example - we support different groups, whom (via role access) should only have access to certain data.  IE.... we wouldn't want let userA and userB both have access to these index patterns, as we don't want to allow them access to each other's data.

**Comment by WassimDhib** [comment]
Hi


Same problem here : https://discuss.opendistrocommunity.dev/t/trace-analysis-and-multi-tenant/4602/2
Trace Analysis plugin require `indices:data/read/search` permission on all `otel-v1-apm-span-*` indices
so either you user have permissions for all indices, or nothing works

I think that supporting index patterns for wildcard indices is the solution.

**Comment by ryn9** [comment]
@anirudha 
>>Currently we don't support index patterns for wildcard indices, this will be a supported in the next few releases. The wildcards will match any permission-ed index available to user for this roles.

Does that mean the plugins would also be moving away from requiring specific index names?
Perhaps create different 'views' that, upon creation, could use different index patterns?

**Comment by ryn9** [comment]
@anirudha @eugenesk24 - wondering if you folks have confirmed index pattern support for PPL for 1.3 yet.
It doesn't answer all of the questions above, but I believe is an important step for PPL adoption.

**Comment by eugenesk24** [comment]
Sorry for the late reply @ryn9 ! We have added multiple index use through [comma-separate indices](https://github.com/opensearch-project/sql/issues/396) and wildcard matching.

**Comment by ryn9** [comment]
@eugenesk24 .. awesome. I am excited to see it :)

**Comment by eugenesk24** [comment]
Merged to main for 1.3 release
